### DISCUSSION TITLE
Prisma Client Extension not called for nested creates?

### DISCUSSION THREAD

**Discussion by Tobjoern** [prisma/prisma#20554] [Q&A] [answered]
### Question

I'm currently building a chat-app, which features `conversations` and `conversationParts` (i.e. messages etc.).
My goal is to create a property `lastConvoPartSentAtMillis` on a conversation (it should be an actualy property, for performance reasons), which is updated to the millis, when the last message was sent automatically, I intend to use client extensions for that.

My initial idea was, to add queries to `create` and `createMany`, for the `conversationPart` object, and then update the conversation, i.e.:

```ts
async create({ model, operation, args, query }) {
	const result = await query(args);

	// am I allowed to do this?
	await prisma.conversation.update({
		where: {
			id: result.conversationId,
		},
		data: {
			lastConvoPartSentAtMillis: result.createdAt
				? result.createdAt
				: new Date(),
		},
	});

	return result;
},
async createMany({ model, operation, args, query }) {
// omitted for brevity, but does the same thing as create, but for 'createMany'
```

The problem is, when creating a conversation, with nested `ConversationPart` creates, those methods are not called at all.
```ts
const createdConversation = await prisma.conversation.create({
  data: {
    Team: {
      connect: { id: team.id },
    },
    ConversationPart: {
      createMany: {
        data: createdMessages.map((message) => {
          return {
            type: ConversationPartType.MESSAGE,
            conversationPartMessageId: message.id,
            createdAt: message.date,
          };
        }),
      },
    },
  },
});
```

Is this a bug, or does it work as intended? Also, what would be the ideal approach to achieve, what I'm trying?

**Comment by nurul3101** [comment, marked as answer]
Hello @Tobjoern 👋 

Based on this [reference](https://www.prisma.io/docs/concepts/components/prisma-client/custom-validation#input-validation-with-prisma-client-extensions), this behaviour is expected.
<img width="696" alt="Screenshot 2023-08-04 at 6 04 39 PM" src="https://github.com/prisma/prisma/assets/13049676/3e230abb-5d70-4bfd-b8a2-338d1318ecf7">

As a workaround, after creating the `conversation` and its `conversationPart`, you can immediately update the `lastConvoPartSentAtMillis` property of the conversation model.

It could look something like this:
```ts
const createdConversation = await prisma.conversation.create({
  data: {
    Team: {
      connect: { id: team.id },
    },
    ConversationPart: {
      createMany: {
        data: createdMessages.map((message) => {
          return {
            type: ConversationPartType.MESSAGE,
            conversationPartMessageId: message.id,
            createdAt: message.date,
          };
        }),
      },
    },
  },
});

// After creating the conversation and its parts, update the conversation
const lastMessageDate = createdMessages[createdMessages.length - 1].date;
await prisma.conversation.update({
  where: {
    id: createdConversation.id,
  },
  data: {
    lastConvoPartSentAtMillis: lastMessageDate,
  },
});
```
You can also wrap this logic in [interactive transactions](https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide#interactive-transactions) to ensure that `conversation` model gets correctly updated.

**Reply by Tobjoern** [reply]
Hey @nurul3101 thanks a lot for the quick response!
I primarily decided on trying out client extensions, to make the code more understandable + maintainable (i.e. so I don't have to remember to set the lastConvoPartSentAtMillis everytime I create a new message).

Do you know, if there are any plans to solve this problem? I think my usecase isn't uncommon at all.

**Reply by nurul3101** [reply]
Would you mind creating a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for this use case?

This way, you would be notified as soon as we have an update. 🙏

**Reply by vleandersson** [reply]
I think this feature can be really neat, especially as it would potentially allow Prisma to create patched temporal tables for Postgresql 🙏 I believe nesting is also not available for models, which practically makes it a global middleware interface 🤔

**Reply by JerryBels** [reply]
Just got blocked by that as well. I wanted to denormalize some data, adding to a count each time some number is inserted in a specific model, but if nested aren't taken into account then I can't guarantee it will always be working. Back to manual adding everywhere :/

**Reply by GabrielRMCorrea** [reply]
Prisma Extensions not working on nesting operations is a big turn off, it makes them useless
